Elasticsearch storage driver

@hacpai 
Hi I'm waiting for a long time for this driver, and I'm started to test it right away and found some critical issues. 
Elasticsearche storage driver works OK, when sending data to default url "http://localhost:9200", but when i trying to send to remote elasticsearch I'm getting error. 

When I'm trying to write to the existing Elasticsearch:
```
core@77cc656c-1adc-473c-ae2a-82fb3c836ede ~/bin $ ncat --telnet 10.128.129.121 9200
### everything is ok.
core@77cc656c-1adc-473c-ae2a-82fb3c836ede ~/bin $ bin/cadvisor -storage_driver="elasticsearch" -port=8099 -alsologtostderr=true -storage_driver_es_host="http://10.128.129.121:9200"
panic: no Elasticsearch node available

goroutine 1 [running]:
github.com/google/cadvisor/storage/elasticsearch.New(0xc208102780, 0x24, 0xb66370, 0x8, 0xb45230, 0x5, 0x7ffd14dc5da7, 0x1a, 0x0, 0x0, ...)
	/home/tpost/go/src/github.com/google/cadvisor/storage/elasticsearch/elasticsearch.go:104 +0x1a6
main.NewMemoryStorage(0x7ffd14dc5d60, 0xd, 0xbc1830, 0x0, 0x0)
	/home/tpost/go/src/github.com/google/cadvisor/storagedriver.go:111 +0x716
main.main()
	/home/tpost/go/src/github.com/google/cadvisor/cadvisor.go:63 +0x18a

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x11ffe80)
	/home/tpost/go/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:839 +0x78
created by github.com/golang/glog.init·1
	/home/tpost/go/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:406 +0x2a7

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 14 [IO wait]:
net.(*pollDesc).Wait(0xc208011790, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208011790, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208011730, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x7fe991287fb0, 0xc208080760)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20803a2d0, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
net/http.noteEOFReader.Read(0x7fe99128a278, 0xc20803a2d0, 0xc20807b918, 0xc20800f000, 0x1000, 0x1000, 0x9ba7c0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc20801fe00, 0xc20800f000, 0x1000, 0x1000, 0xc208012000, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc208038b40)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc208038b40, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc20807b8c0)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 11 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 15 [select]:
net/http.(*persistConn).writeLoop(0xc20807b8c0)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc
```

When I'm trying to write to unexciting elasticsearch:
```
core@77cc656c-1adc-473c-ae2a-82fb3c836ede ~/bin $ ncat --telnet 10.128.129.12 9200          
Ncat: Connection refused.


core@77cc656c-1adc-473c-ae2a-82fb3c836ede ~/bin $ bin/cadvisor -storage_driver="elasticsearch" -port=8099 -alsologtostderr=true -storage_driver_es_host="http://10.128.129.12:9200"
panic: no Elasticsearch node available

goroutine 1 [running]:
github.com/google/cadvisor/storage/elasticsearch.New(0xc208102780, 0x24, 0xb66370, 0x8, 0xb45230, 0x5, 0x7ffec742bda8, 0x19, 0x0, 0x0, ...)
	/home/tpost/go/src/github.com/google/cadvisor/storage/elasticsearch/elasticsearch.go:104 +0x1a6
main.NewMemoryStorage(0x7ffec742bd61, 0xd, 0xbc1830, 0x0, 0x0)
	/home/tpost/go/src/github.com/google/cadvisor/storagedriver.go:111 +0x716
main.main()
	/home/tpost/go/src/github.com/google/cadvisor/cadvisor.go:63 +0x18a

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x11ffe80)
	/home/tpost/go/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:839 +0x78
created by github.com/golang/glog.init·1
	/home/tpost/go/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:406 +0x2a7

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 11 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35
```

Also I have a reason to believe, that no matter which Elasticsearch URL I set, application still trying to write to the Default URL(http://localhost:9200). 

I'm using CoreOS. And everything is works when I'm running Cadvisor on the machine with running local Elasticsearch container.

@hacpai 
I have no idea where problem is. Here is very strange behavior.

Running:
```
core@8d3c56f4-3468-4c69-9afb-6202099641f0 ~/bin $ curl 127.0.0.1:9200
{
  "status" : 200,
  "name" : "Black Tarantula",
  "cluster_name" : "elasticsearch",
  "version" : {
    "number" : "1.7.1",
    "build_hash" : "b88f43fc40b0bcd7f173a1f9ee2e97816de80b19",
    "build_timestamp" : "2015-07-29T09:54:16Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.4"
  },
  "tagline" : "You Know, for Search"
}

core@8d3c56f4-3468-4c69-9afb-6202099641f0 ~/bin $ curl http://10.128.129.121:9200/
{
  "status" : 200,
  "name" : "Black Tarantula",
  "cluster_name" : "elasticsearch",
  "version" : {
    "number" : "1.7.1",
    "build_hash" : "b88f43fc40b0bcd7f173a1f9ee2e97816de80b19",
    "build_timestamp" : "2015-07-29T09:54:16Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.4"
  },
  "tagline" : "You Know, for Search"
}

core@8d3c56f4-3468-4c69-9afb-6202099641f0 ~/bin $ curl http://remote_elk:9200/
{
  "status" : 200,
  "name" : "Ecstasy",
  "cluster_name" : "elasticsearch",
  "version" : {
    "number" : "1.7.1",
    "build_hash" : "b88f43fc40b0bcd7f173a1f9ee2e97816de80b19",
    "build_timestamp" : "2015-07-29T09:54:16Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.4"
  },
  "tagline" : "You Know, for Search"
}
```

As you can see here is:
127.0.0.1 its localhost for **10.128.129.121** and **remote_elk** its some remote elk.

Starting cadvisor on the **10.128.129.121** CoreOS:
```
core@8d3c56f4-3468-4c69-9afb-6202099641f0 ~/bin $ ./cadvisor -storage_driver="elasticsearch" -port=8099 -alsologtostderr=true -storage_driver_es_host="http://remote_elk:9200"
panic: no Elasticsearch node available

goroutine 1 [running]:
github.com/google/cadvisor/storage/elasticsearch.New(0xc2080a4780, 0x24, 0xb663d0, 0x8, 0xb45290, 0x5, 0x7ffdf2ef8da9, 0x24, 0x0, 0x0, ...)
	/home/tpost/go/src/github.com/google/cadvisor/storage/elasticsearch/elasticsearch.go:104 +0x1af
main.NewMemoryStorage(0x7ffdf2ef8d62, 0xd, 0xbc1890, 0x0, 0x0)
	/home/tpost/go/src/github.com/google/cadvisor/storagedriver.go:111 +0x716
main.main()
	/home/tpost/go/src/github.com/google/cadvisor/cadvisor.go:63 +0x18a

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x11ffe80)
	/home/tpost/go/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:839 +0x78
created by github.com/golang/glog.init·1
	/home/tpost/go/src/github.com/google/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:406 +0x2a7

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 11 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 15 [IO wait]:
net.(*pollDesc).Wait(0xc208011480, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc208011480, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208011420, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x7f4ea0dc8fb0, 0xc2080a0ab0)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20803a2d8, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
net/http.noteEOFReader.Read(0x7f4ea0dcb278, 0xc20803a2d8, 0xc2080fd658, 0xc20800f000, 0x1000, 0x1000, 0x9ba820, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc20801fe60, 0xc20800f000, 0x1000, 0x1000, 0xc208012000, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc208038c60)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc208038c60, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc2080fd600)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 16 [select]:
net/http.(*persistConn).writeLoop(0xc2080fd600)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc
```

now trying local elk
```
core@8d3c56f4-3468-4c69-9afb-6202099641f0 ~/bin $ ./cadvisor -storage_driver="elasticsearch" -port=8099 -alsologtostderr=true -storage_driver_es_host="http://10.128.129.121:9200"
Elasticsearch returned with code 200 and version 1.7.1I0910 15:57:04.568553 28608 storagedriver.go:128] Using backend storage type "elasticsearch"
```

and the same but with 127.0.0.1
```
core@8d3c56f4-3468-4c69-9afb-6202099641f0 ~/bin $ ./cadvisor -storage_driver="elasticsearch" -port=8099 -alsologtostderr=true -storage_driver_es_host="http://127.0.0.1:9200"
Elasticsearch returned with code 200 and version 1.7.1I0910 15:57:58.997353 29003 storagedriver.go:128] Using backend storage type "elasticsearch"
```


If I run any of this commands on the other CoreOS machine(where is no local elk installed) all 3 commands will fail.

@hacpai 
Still can't write to remote elasticsearch :(

```
core@38a63250-c3e0-4328-b08f-e23fdb6029e4 ~/bin $ ./cadvisor -storage_driver="elasticsearch" -storage_driver_es_host="http://10.128.129.121:9200"
panic: no Elasticsearch node available

goroutine 1 [running]:
github.com/google/cadvisor/storage/elasticsearch.New(0xc2080a4810, 0x24, 0xb66830, 0x8, 0xb456f0, 0x5, 0x7ffdfdf97db3, 0x1a, 0x0, 0x0, ...)
	/home/tpost/go/src/github.com/google/cadvisor/storage/elasticsearch/elasticsearch.go:104 +0x1af
main.NewMemoryStorage(0x7ffdfdf97d8d, 0xd, 0xbc1db0, 0x0, 0x0)
	/home/tpost/go/src/github.com/hacpai/cadvisor/storagedriver.go:111 +0x716
main.main()
	/home/tpost/go/src/github.com/hacpai/cadvisor/cadvisor.go:63 +0x18a

goroutine 5 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x1200e80)
	/home/tpost/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:839 +0x78
created by github.com/golang/glog.init·1
	/home/tpost/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:406 +0x2a7

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 14 [IO wait]:
net.(*pollDesc).Wait(0xc2080116b0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:84 +0x47
net.(*pollDesc).WaitRead(0xc2080116b0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:89 +0x43
net.(*netFD).Read(0xc208011650, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x7f5093909fb0, 0xc2080a0be0)
	/usr/local/go/src/net/fd_unix.go:242 +0x40f
net.(*conn).Read(0xc20803a2d0, 0xc20800f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:121 +0xdc
net/http.noteEOFReader.Read(0x7f509390c208, 0xc20803a2d0, 0xc2080fd0d8, 0xc20800f000, 0x1000, 0x1000, 0x9bac20, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1270 +0x6e
net/http.(*noteEOFReader).Read(0xc20801fec0, 0xc20800f000, 0x1000, 0x1000, 0xc208012000, 0x0, 0x0)
	<autogenerated>:125 +0xd4
bufio.(*Reader).fill(0xc208038b40)
	/usr/local/go/src/bufio/bufio.go:97 +0x1ce
bufio.(*Reader).Peek(0xc208038b40, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xf0
net/http.(*persistConn).readLoop(0xc2080fd080)
	/usr/local/go/src/net/http/transport.go:842 +0xa4
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:660 +0xc9f

goroutine 11 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:21 +0x1f
created by os/signal.init·1
	/usr/local/go/src/os/signal/signal_unix.go:27 +0x35

goroutine 15 [select]:
net/http.(*persistConn).writeLoop(0xc2080fd080)
	/usr/local/go/src/net/http/transport.go:945 +0x41d
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:661 +0xcbc
core@38a63250-c3e0-4328-b08f-e23fdb6029e4 ~/bin $ curl 10.128.129.121:9200
{
  "status" : 200,
  "name" : "Hack",
  "cluster_name" : "elasticsearch",
  "version" : {
    "number" : "1.7.1",
    "build_hash" : "b88f43fc40b0bcd7f173a1f9ee2e97816de80b19",
    "build_timestamp" : "2015-07-29T09:54:16Z",
    "build_snapshot" : false,
    "lucene_version" : "4.10.4"
  },
  "tagline" : "You Know, for Search"
}
```

PS: I'm not using cadvisor from container currently I'm running binary directly from CoreOS.

Did you try to connect cadvisor to remote elk?

fixbug-issue-881

cadvisor trying connect to http://127.0.0.1:9200/ just ignoring provided -storage_driver_es_host="http://192.168.0.101:9200"



root@98fde4963846:/go/src/github.com/hacpai/cadvisor# ./cadvisor -storage_driver="elasticsearch" -port=8099 -alsologtostderr=true -storage_driver_es_host="http://192.168.0.101:9200"
panic: Get http://127.0.0.1:9200/: dial tcp 127.0.0.1:9200: getsockopt: connection refused

goroutine 1 [running]:
github.com/google/cadvisor/storage/elasticsearch.New(0xc820121510, 0xc, 0xbb91a0, 0x8, 0xba8f68, 0x5, 0x7ffef34ffe61, 0x19, 0x0, 0x0, ...)
	/go/src/github.com/google/cadvisor/storage/elasticsearch/elasticsearch.go:111 +0x261
main.NewMemoryStorage(0x7ffef34ffe1a, 0xd, 0xc820126580, 0x0, 0x0)
	/go/src/github.com/hacpai/cadvisor/storagedriver.go:111 +0xe86
main.main()
	/go/src/github.com/hacpai/cadvisor/cadvisor.go:63 +0x193

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 15 [select]:
net/http.(*persistConn).writeLoop(0xc8200b5970)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 6 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37

goroutine 7 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x1276220)
	/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:839 +0x67
created by github.com/golang/glog.init.1
	/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/github.com/golang/glog/glog.go:406 +0x297

goroutine 14 [IO wait]:
net.runtime_pollWait(0x7fa6b1356790, 0x72, 0xc820010170)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc820123020, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc820123020, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc820122fc0, 0xc820115000, 0x1000, 0x1000, 0x0, 0x7fa6b1350050, 0xc820010170)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc8200263c0, 0xc820115000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.noteEOFReader.Read(0x7fa6b1356850, 0xc8200263c0, 0xc8200b59c8, 0xc820115000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc82011b740, 0xc820115000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc8200158c0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc8200158c0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8200b5970)
	/usr/local/go/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 24 [select]:
gopkg.in/olivere/elastic%2ev2.(*Client).sniffer(0xc8200739e0)
	/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/gopkg.in/olivere/elastic.v2/client.go:523 +0x1af
created by gopkg.in/olivere/elastic%2ev2.NewClient
	/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/gopkg.in/olivere/elastic.v2/client.go:223 +0x5bf

goroutine 19 [IO wait]:
net.runtime_pollWait(0x7fa6b13566d0, 0x72, 0xc820010170)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8201234f0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8201234f0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc820123490, 0xc820131000, 0x1000, 0x1000, 0x0, 0x7fa6b1350050, 0xc820010170)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820026408, 0xc820131000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.noteEOFReader.Read(0x7fa6b1356850, 0xc820026408, 0xc8200b5a78, 0xc820131000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc82011bec0, 0xc820131000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc820015e60)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc820015e60, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8200b5a20)
	/usr/local/go/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 20 [select]:
net/http.(*persistConn).writeLoop(0xc8200b5a20)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 22 [IO wait]:
net.runtime_pollWait(0x7fa6b1356610, 0x72, 0xc820010170)
	/usr/local/go/src/runtime/netpoll.go:157 +0x60
net.(*pollDesc).Wait(0xc8201236b0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8201236b0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc820123650, 0xc820147000, 0x1000, 0x1000, 0x0, 0x7fa6b1350050, 0xc820010170)
	/usr/local/go/src/net/fd_unix.go:232 +0x23a
net.(*conn).Read(0xc820026420, 0xc820147000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
net/http.noteEOFReader.Read(0x7fa6b1356850, 0xc820026420, 0xc8200b5b28, 0xc820147000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/http/transport.go:1370 +0x67
net/http.(*noteEOFReader).Read(0xc82014e000, 0xc820147000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xd0
bufio.(*Reader).fill(0xc82014c2a0)
	/usr/local/go/src/bufio/bufio.go:97 +0x1e9
bufio.(*Reader).Peek(0xc82014c2a0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/bufio/bufio.go:132 +0xcc
net/http.(*persistConn).readLoop(0xc8200b5ad0)
	/usr/local/go/src/net/http/transport.go:876 +0xf7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 23 [select]:
net/http.(*persistConn).writeLoop(0xc8200b5ad0)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d

goroutine 25 [select]:
gopkg.in/olivere/elastic%2ev2.(*Client).healthchecker(0xc8200739e0)
	/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/gopkg.in/olivere/elastic.v2/client.go:689 +0x1b4
created by gopkg.in/olivere/elastic%2ev2.NewClient
	/go/src/github.com/hacpai/cadvisor/Godeps/_workspace/src/gopkg.in/olivere/elastic.v2/client.go:224 +0x5e1

--- a/storage/elasticsearch/elasticsearch.go
+++ b/storage/elasticsearch/elasticsearch.go
@@ -17,6 +17,7 @@ package elasticsearch
 import (
 	"fmt"
 	"sync"
+	"time"
 
 	info "github.com/google/cadvisor/info/v1"
 	storage "github.com/google/cadvisor/storage"
@@ -67,7 +68,7 @@ func (self *elasticStorage) AddStats(ref info.ContainerReference, stats *info.Co
 		// Add some default params based on ContainerStats
 		detail := self.containerStatsAndDefaultValues(ref, stats)
 		// Index a cadvisor (using JSON serialization)
-		put, err := self.client.Index().
+		_, err := self.client.Index().
 			Index(self.indexName).
 			Type(self.typeName).
 			BodyJson(detail).
@@ -76,7 +77,6 @@ func (self *elasticStorage) AddStats(ref info.ContainerReference, stats *info.Co
 			// Handle error
 			panic(fmt.Errorf("failed to write stats to ElasticSearch- %s", err))
 		}
-		fmt.Printf("Indexed tweet %s to index %s, type %s\n", put.Id, put.Index, put.Type)
 	}()
 	return nil
 }
@@ -93,19 +93,24 @@ func New(machineName,
 	indexName,
 	typeName,
 	elasticHost string,
+	enableSniffer bool,
 ) (storage.StorageDriver, error) {
 	// Obtain a client and connect to the default Elasticsearch installation
 	// on 127.0.0.1:9200. Of course you can configure your client to connect
 	// to other hosts and configure it in various other ways.
 	client, err := elastic.NewClient(
-		elastic.SetURL(elasticHost))
+		elastic.SetHealthcheck(true),
+		elastic.SetSniff(enableSniffer),
+		elastic.SetHealthcheckInterval(30*time.Second),
+		elastic.SetURL(elasticHost),
+	)
 	if err != nil {
 		// Handle error
 		panic(err)
 	}
 
 	// Ping the Elasticsearch server to get e.g. the version number
-	info, code, err := client.Ping().Do()
+	info, code, err := client.Ping().URL(elasticHost).Do()
 	if err != nil {
 		// Handle error
 		panic(err)

--- a/storagedriver.go
+++ b/storagedriver.go
@@ -39,9 +39,10 @@ var argDbTable = flag.String("storage_driver_table", "stats", "table name")
 var argDbIsSecure = flag.Bool("storage_driver_secure", false, "use secure connection with database")
 var argDbBufferDuration = flag.Duration("storage_driver_buffer_duration", 60*time.Second, "Writes in the storage driver will be buffered for this duration, and committed to the non memory backends as a single transaction")
 var storageDuration = flag.Duration("storage_duration", 2*time.Minute, "How long to keep data stored (Default: 2min).")
-var argElasticHost = flag.String("storage_driver_es_host", "http://localhost:9200", "database host:port")
-var argIndexName = flag.String("storage_driver_index", "cadvisor", "index name")
-var argTypeName = flag.String("storage_driver_type", "stats", "type name")
+var argElasticHost = flag.String("storage_driver_es_host", "http://localhost:9200", "ElasticSearch host:port")
+var argIndexName = flag.String("storage_driver_es_index", "cadvisor", "ElasticSearch index name")
+var argTypeName = flag.String("storage_driver_es_type", "stats", "ElasticSearch type name")
+var argEnableSniffer = flag.Bool("storage_driver_es_enable_sniffer", false, "ElasticSearch uses a sniffing process to find all nodes of your cluster by default, automatically")
 
 // Creates a memory storage with an optional backend storage option.
 func NewMemoryStorage(backendStorageName string) (*memory.InMemoryCache, error) {
@@ -108,6 +109,7 @@ func NewMemoryStorage(backendStorageName string) (*memory.InMemoryCache, error) 
 			*argIndexName,
 			*argTypeName,
 			*argElasticHost,
+			*argEnableSniffer,
 		)
 	case "statsd":
 		backendStorage, err = statsd.New(

